function to process SpliceAI scores

KoalaQin opened the pull request:
Similar to Pangolin, we take a maximum score across 4 types of splice consequences, then a maximum across genes if there are multiple genes. This is processed from Illumina's precomputed scores for SNVs and Indels, and new indels shown in v4 exomes.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+                consequences[ht.splice_ai.delta_scores.index(ht.splice_ai.ds_max)],
+                "no_consequence",
+            ),
+            position_max=hl.if_else(
```
```suggestion
            ds_max_position=hl.if_else(
```

mike-w-wilson replied:
Same deal as above

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    )
+    ht = ht.annotate(
+        splice_ai=ht.splice_ai.annotate(
+            consequence_max=hl.if_else(
```
I think this is a bit more information but no strong feeling.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+        staging=hl.agg.take(
+            hl.struct(
+                ds_max=ht.splice_ai.ds_max,
+                position_max=ht.splice_ai.position_max,
+                consequence_max=ht.splice_ai.consequence_max,
```
```suggestion
                ds_max_position=ht.splice_ai.ds_max_position,
                ds_max_consequence=ht.splice_ai.ds_max_consequence,
```
Same as above.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    # Annotate info.max_DS with the max of DS_AG, DS_AL, DS_DG, DS_DL in info.
+    # delta_score array is |DS_AG|DS_AL|DS_DG|DS_DL
+    logger.info(
+        "Getting the max SpliceAI score for each variant across consequences per"
+        " gene..."
```
```suggestion
        "Getting the max SpliceAI score across consequences for each variant per"
         "gene..."
```

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+        )
+    )
+
+    logger.info("Getting the max SpliceAI score for each variant across genes...")
```
```suggestion
    logger.info("Getting the max SpliceAI score at each position across genes...")
```
I think this is more accurate since you are grouping by locus?

KoalaQin replied:
grouping by both locus and alleles, it's not just by position.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+
+    logger.info("Getting the max SpliceAI score for each variant across genes...")
+    ht2 = ht.group_by(*ht.key).aggregate(
+        staging=hl.agg.take(
```
Cool, I didnt know this existed!

KoalaQin replied:
I asked in Zulip, Konrad told me that.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    )
+
+    logger.info("Getting the max SpliceAI score for each variant across genes...")
+    ht2 = ht.group_by(*ht.key).aggregate(
```
Do we need to create ht2? Can we just reassign the ht? Or does that not work?

KoalaQin replied:
I think I tried, it didn't work, or I get repeated max.

mike-w-wilson reviewed (changes requested):
Back to you! Mostly just documentation and a bit on organization.

KoalaQin commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    all SNVs precomputed: spliceai_scores.masked.snv.hg38.vcf.bgz, downloaded from https://basespace.illumina.com/s/5u6ThOblecrh
+    all indels precomputed: spliceai_scores.masked.indel.hg38.vcf.bgz, downloaded from https://basespace.illumina.com/s/5u6ThOblecrh
+    new indels from gnomAD v4: gnomad_v4_new_indels.spliceai_masked.vcf.bgz, requested Illumina to compute for us.
+    We may have lost the SpliceAI score for gnomAD v3 indels, which makes it hard to process the raw scores to get a maximum and retain the position information.
+    Before we didn't `explode` the SpliceAI scores, so we only had scores for the first gene for each variant.
```
We lost that one, Illumina sent back by email but the link is expired and William can't find it. I also can't find any trace from his code.

ch-kr commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -107,6 +231,25 @@ def main(args):
 
+    if args.spliceai:
+        logger.info("Creating SpliceAI Hail Table for GRCh38...")
+        snvs_path = (
+            "gs://gnomad-qin/v4_annotations/spliceai_scores.masked.snv.hg38.vcf.bgz"
+        )
+        indels_path = (
+            "gs://gnomad-qin/v4_annotations/spliceai_scores.masked.indel.hg38.vcf.bgz"
+        )
+        new_indels_path = "gs://gnomad-qin/v4_annotations/gnomad_v4_new_indels.spliceai_masked.vcf.bgz"
+        header_file_path = "gs://gnomad-qin/v4_annotations/spliceai.vcf.header"
```
I think we should keep this full resource spliceAI resource internal given the discussion around licensing earlier this year (copy the current files into a new bucket with autoclass enabled).

for the other resources, we shouldn't need to copy the import files, so something like
```
cadd = VersionedTableResource(
    versions={
        "snvs": GnomadPublicTableResource(
            path="gs://gcp-public-data--gnomad/resources/grch38/in_silico_predictors/CADD_v1.6_SNVs.ht",
        ),
        "indels3.0": GnomadPublicTableResource(
            path="gs://gnomad-public-requester-pays/resources/grch38/cadd/cadd-indels-gnomad.3.0.ht",
            import_func=_load_cadd_raw,
            import_args={
                "path":""gs://gnomad-qin/v4_annotations/cadd.v1.6.gnomad.genomes.r3.0.indel.tsv.bgz"
            }
        ),
    }
)
```

mike-w-wilson replied:
So I agree with KC above except for the import_args path.  I don't think any files used in production should live in personal buckets. When people clean their personal buckets up is unpredictable and I don't think we should have to dig into the codebase to find a source file. Having the file in an organized and well labeled bucket is nice. It shouldnt be public but it can live in  `gs://gnomad/v4.0/annotations/in_silico_predictors` which is an intuitive, existing bucket.

KoalaQin replied:
We created a new bucket for the nonpublic raw data: gs://gnomad-insilico, I'm moving them there and adding to gnomad/resources/grch38/reference_data.py, I will create another PR in gnomad_methods.

KoalaQin replied:
Aabout copying resources to public bucket, for dbSNP, I just did as before, https://github.com/broadinstitute/gnomad_methods/blob/12a11ae74d09ed46e92a2e3eb49182dc05aaa4b8/gnomad/resources/grch38/reference_data.py#L223, but it doesn’t make sense to me to keep both raw data & HT in the public bucket, for the other insilico predictors, I’m going to import from gnomad-insilico instead of from gcp-public-data--gnomad.

KoalaQin replied:
I think I know why now, it has to be imported from the public buckets: 
`Job failed with message [ValueError: GnomadPublicResource requires a path to a file in one of the public gnomAD buckets (gnomad-public, gnomad-public-requester-pays)]. Additional details can be found at:`

KoalaQin commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    ht = mt.rows()
+
+    logger.info("Exploding SpliceAI scores...")
+    # `explode` will eliminate rows with empty array, but the varaints with multiple genes will extend the number of rows.
```
this is because SpliceAI only has predictions for variants inside the gene body, so number of lines != number of variants with score(s), I will put a note here.

KoalaQin commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +89,130 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    ht = ht.annotate(
+        splice_ai=ht.splice_ai.annotate(ds_max=hl.max(ht.splice_ai.delta_scores))
+    )
+    ht = ht.annotate(
```
they don't want to show the position in the browser, so we can delete this after the commit.

mike-w-wilson replied:
We can just remove it now since we wont be using it.

KoalaQin commented:
I addressed your comments, but I'm thinking maybe we can just merge these scores to one HT for both v4 genomes and exomes, now that we are making the resources (except SpliceAI). This code is just to process the scores, do we really need to keep the processed big HT or just for our gnomAD variants?

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
 
 
+def create_spliceai_grch38_ht(
+    snvs_path: str, indels_path: str, new_indels_path: str, header_file_path: str
```
I think we should be consistent with how we handle paths in the same module. CADD's import has all paths inside the function. Lets move spliceAIs paths into the function as well and remove the parameters so in main youd just call `create_spliceai_grch38_ht()`

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    :param str new_indels_path: Path to the gnomAD v4 indels.
+    :return: Hail Table with SpliceAI scores for GRCh38.
+    """
+
```
You could move the paths here.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    ht = ht.explode(ht.info.SpliceAI)
+
+    logger.info("Annotating SpliceAI scores...")
+    # there will only be one gene in the array after exploding
+    gene_symbol = ht.info.SpliceAI.split(delim="\\|")[1:2][0]
```
```suggestion
```
This isn't used in the logic for the return field and is removed by the final select so lets remove it.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    gene_symbol = ht.info.SpliceAI.split(delim="\\|")[1:2][0]
+    # delta_score array for 4 splicing consequences: DS_AG|DS_AL|DS_DG|DS_DL
+    delta_scores = ht.info.SpliceAI.split(delim="\\|")[2:6]
+    # position array is |AG|AL|DG|DL
+    positions = ht.info.SpliceAI.split(delim="\\|")[6:10]
```
```suggestion
```
This isn't used in the logic for the return field and is removed by the final select so lets remove it.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    ht = ht.annotate(
+        splice_ai=ht.splice_ai.annotate(ds_max=hl.max(ht.splice_ai.delta_scores))
+    )
+    ht = ht.annotate(
+        splice_ai=ht.splice_ai.annotate(
+            ds_max_consequence=hl.if_else(
+                ht.splice_ai.ds_max > 0,
+                consequences[ht.splice_ai.delta_scores.index(ht.splice_ai.ds_max)],
+                "no_consequence",
+            ),
+            ds_max_position=hl.if_else(
+                ht.splice_ai.ds_max > 0,
+                ht.splice_ai.positions[
+                    ht.splice_ai.delta_scores.index(ht.splice_ai.ds_max)
+                ],
+                hl.missing(hl.tint32),
+            ),
+        )
+    )
+
```
```suggestion
```

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    delta_scores = ht.info.SpliceAI.split(delim="\\|")[2:6]
+    # position array is |AG|AL|DG|DL
+    positions = ht.info.SpliceAI.split(delim="\\|")[6:10]
+    ht = ht.annotate(
+        splice_ai=hl.struct(
+            gene_symbol=gene_symbol,
+            delta_scores=hl.map(lambda x: hl.float32(x), delta_scores),
+            positions=hl.map(lambda x: hl.int32(x), positions),
+        )
+    )
+
+    # Annotate info.max_DS with the max of DS_AG, DS_AL, DS_DG, DS_DL in info.
+    logger.info(
+        "Getting the max SpliceAI score across consequences for each variant pergene..."
+    )
+    consequences = hl.literal(
+        ["acceptor_gain", "acceptor_loss", "donor_gain", "donor_loss"]
+    )
+    ht = ht.annotate(
+        splice_ai=ht.splice_ai.annotate(ds_max=hl.max(ht.splice_ai.delta_scores))
+    )
```
```suggestion
    ht = ht.annotate(delta_scores=hl.map(lambda x: hl.float32(x), delta_scores))
     # Annotate ds_max with the max of DS_AG, DS_AL, DS_DG, DS_DL in info.
    logger.info(
        "Getting the max SpliceAI score across consequences for each variant per"
        " gene..."
    )
    ht = ht.select(ds_max=hl.max(ht.delta_scores))
    )
```
I don't think you need the struct just yet since we arent using those fields and it's less easier to handle the single top level annotation in the code rather than a single nested annotation.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    ht = spliceai_snvs.union(spliceai_indels).union(spliceai_new_indels)
+
+    logger.info("Exploding SpliceAI scores...")
+    # `explode` will eliminate rows with empty array, some variants don't have
+    # score might because they are located inside gene body, but the variants
+    # fall on multiple genes will extend the number of rows.
```
```suggestion
    # `explode` because some variants fall on multiple genes and have a score per gene.
    # All rows without a SpliceAI score, an empty array, are removed through explode.
```

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,131 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    )
+
+    logger.info("Getting the max SpliceAI score for each variant across genes...")
+    # The block of code below keeps all the info for the variant with the max
+    # SpliceAI score: ds_max, ds_max_consequence, ds_max_position, gene_symbol.
+    # We want to keep it here in case we want to use it in the future.
+    ht2 = ht.group_by(*ht.key).aggregate(
+        staging=hl.agg.take(
+            hl.struct(
+                ds_max=ht.splice_ai.ds_max,
+                ds_max_position=ht.splice_ai.ds_max_position,
+                ds_max_consequence=ht.splice_ai.ds_max_consequence,
+                gene=ht.splice_ai.gene_symbol,
+            ),
+            1,
+            ordering=-ht.splice_ai.ds_max,
+        )
+    )
+
+    logger.info("Annotating SpliceAI scores in right format...")
+    # `aggregate` put everything in an array, we need to extract the struct
+    # from the array.
+    # For the v4 release, we will only keep the max SpliceAI score.
+    ht2 = ht2.annotate(
+        splice_ai=hl.struct(
+            ds_max=hl.float32(ht2.staging.ds_max[0]),
+        )
+    )
+
+    ht2 = ht2.select("splice_ai")
+    return ht2
```
```suggestion
    ht = ht.collect_by_key()
    ht = ht.select(splice_ai=hl.struct(ds_max=hl.max(ht.values.ds_max)))

    return ht
```
Also removing the no longer needed fields and using a collect_by_ket in place of group_by here. This reduces the amount of code and variables.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -107,6 +233,25 @@ def main(args):
 
+    if args.spliceai:
+        logger.info("Creating SpliceAI Hail Table for GRCh38...")
+        snvs_path = (
+            "gs://gnomad-insilico/spliceai/spliceai_scores.masked.snv.hg38.vcf.bgz"
+        )
+        indels_path = (
+            "gs://gnomad-insilico/spliceai/spliceai_scores.masked.indel.hg38.vcf.bgz"
+        )
+        new_indels_path = "gs://gnomad-insilico/spliceai/spliceai_scores.masked.gnomad_v4_new_indels.hg38.vcf.bgz"
+        header_file_path = "gs://gnomad-insilico/spliceai/spliceai.vcf.header"
```
Move these into the create_spliceai_grch38_ht function.

mike-w-wilson reviewed (changes requested):
I made some pretty significant suggestions. This is mainly removing the unused fields that we decided we no longer needed in the project meeting a few weeks ago. I know you had put in a note about keeping the annotations should we decide to use them in the future and I know it was a decent amount of work to write this code but we should aim to have the code be as readable as possible/what is actually run for production. Removing the extra code increases readability but  also minimizes the compute cost. My suggestions just find the max score across all genes and select that. To keep the spliceAI consequence code around, you could create a branch from this branch before removing it via my suggestions or your own approach and reference that branch in the ticket so we know you already did the work should we revisit it in the future -- that what I would do.

mike-w-wilson commented:
I also pulled your branch and tested my changes locally so I can make a commit or a PR to merge changes into your branch if you would prefer that to the github suggestion feature.

KoalaQin commented:
Thanks for going through and testing it. I agree with you, this looks much cleaner. I created a backup branch in case we need to revisit.

KoalaQin commented:
One question: do you think we should just merge the table with v4 genomes and exomes now to save some space or later?

mike-w-wilson commented:
Sorry, I'm unsure what you mean by merge the table? Do you mean put it on the VDSs? If so, we should not do that. We'll add it in the release code.

KoalaQin commented:
I used the vep HT for VRS code, we can use that one if needed, but we can also do it later.

KoalaQin commented:
@mike-w-wilson I'm getting this error in the checks: https://github.com/broadinstitute/gnomad_qc/actions/runs/6110084873/job/16582370828

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,75 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    spliceai_indels = import_spliceai_vcf(indels_path)
+    spliceai_new_indels = import_spliceai_vcf(new_indels_path)
+
+    ht = spliceai_snvs.union(spliceai_indels).union(spliceai_new_indels)
```
So just for consistency, can you do what you did on L88? I'm not sure if I missed this before but for the union on L88, does hail automatically know the second arg is a table and not union's unify arg?

KoalaQin replied:
Yes, `union` is to append rows of multiple HT, I didn't do that because it was `union_rows` for MT.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -64,13 +65,13 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
 
-    snvs = _load_cadd_raw(
-        "gs://gnomad-qin/v4_annotations/cadd.v1.6.whole_genome_SNVs.tsv.bgz"
+    snvs = hl.read_table(
+        "gs://gcp-public-data--gnomad/resources/grch38/in_silico_predictors/CADD_v1.6_SNVs.ht"
```
Does this table still exist here? I can't remember which we removed. However, in general, we should keep the original import around or make a note of how this HT was generated. Without either the note or tsv import, someone who didnt run the code will not know how this HT was created.

KoalaQin replied:
Yes, it exists, we didn't move the ones uploaded to public bucket yet. I can revert it back but change the paths.

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -89,6 +90,75 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
+    # delta_score array for 4 splicing consequences: DS_AG|DS_AL|DS_DG|DS_DL
+    delta_scores = ht.info.SpliceAI.split(delim="\\|")[2:6]
+    ht = ht.annotate(delta_scores=hl.map(lambda x: hl.float32(x), delta_scores))
+    # Annotate ds_max with the max of DS_AG, DS_AL, DS_DG, DS_DL in info.
```
```suggestion
```
I know I rewrote this bit but I think this comment is unneeded.

mike-w-wilson reviewed (commented):
Just a couple more things, suggestions and questions!

mike-w-wilson commented on gnomad_qc/v4/annotations/insilico_predictors.py:
```diff
@@ -64,18 +65,18 @@ def _load_cadd_raw(cadd_tsv) -> hl.Table:
 
-    snvs = _load_cadd_raw(
-        "gs://gnomad-qin/v4_annotations/cadd.v1.6.whole_genome_SNVs.tsv.bgz"
+    snvs = hl.read_table(
```
I think this needs to be _load_raw_cadd now right?

KoalaQin commented:
I'm closing this PR because the rebasing messed up the commits: here is the new PR: https://github.com/broadinstitute/gnomad_qc/pull/423
